### DISCUSSION TITLE
api: plan for `narwhals.stable.v2`

### DISCUSSION THREAD

**Discussion by MarcoGorelli** [narwhals-dev/narwhals#1657] [Announcements]
I think we can aim for `narwhals.stable.v2` some time in 2025

The main things I'd like to achieve are:
- remove "interchange" level of support, and instead only have:
  - "full": must support `Series` and have a defined row order (pandas, polars.DataFrame, pyarrow.Table, modin.DataFrame, cudf.DataFrame)
  - "lazy": must support `collect`, adding/removing/subsetting columns, filtering rows. Initially, at least, row-order-dependent operations (such as `nw.cum_sum`) will be left out. This will include Polars.LazyFrame, DuckDBPyRelation, pyspark.DataFrame, ibis.Table, dask.DataFrame. Later we can add an `order_by` argument to these functions which will become required when working with nw.Lazyframe
- make the breaking changes listed in [breaking changes](https://narwhals-dev.github.io/narwhals/backcompat/#breaking-changes-carried-out-so-far)

Candidate changes:
- rework `from_native`. It will get simpler once we remove interchange-level things, but it's still fairly complicated. Not sure how much I like it to be completely honest, but I also haven't come up with anything better

**Usage of `narwhals.stable.v1` should remain unaffected**

Tentative date: some time in 2025

---

### Why remove support for the dataframe interchange protocol?

- the project has received no updates in over a year, issues and requests go unanswered, and it has no governance model in place so nobody is taking responsibility for it
- only a handful of dataframe libraries implement the protocol, and for most of them, we already provide (or are working to provide) native support in Narwhals. The only exception is Vaex, which itself seems to be deprecated (no releases in ~1.5 years), but we could support that too if there was demand
- from a Narwhals usage perspective, it adds a lot of complexity to have 3 layers of support. It's much simpler to just have: full, and lazy.
---

## Outstanding items

Breaking changes:
- [x] deprecate order-dependent expressions being passed to nw.LazyFrame
- [x] deprecate length-changing order-dependent expressions (Expr.sort, Expr.gather_every, Expr.head, Expr.tail, Expr.sample)
- [x] introduce restrictions on when length-changing expressions can be used in lazy API
- [x] deprecate `eager_or_interchange_only` from `from_native`
- [x] all expressions passed to `group_by(...).agg` should aggregate
- [x] sort out #1780 (to make sure that addressing it doesn't require further API breaks)
- [x] does `nw.LazyFrame` really need `.clone`? #2206
- [x] `native_namespace` -> `backend` in IO functions #1888
- [x] deprecate`LazyFrame.gather_every`
- [x] should `pd.Categorical` get mapped to `nw.Enum`? #2192
- [x] time zones in duckdb #2165
- [x] empty sum: `0` or `null`?
- [x] expression expansion in group-by #2225
- [x] `LazyFrame.with_row_index`: https://github.com/narwhals-dev/narwhals/issues/2307 
- [x] horizontal concatenation #2340
- [ ] ~~IO functions https://github.com/narwhals-dev/narwhals/issues/2116~~
- [x] `utils` public/private: https://github.com/narwhals-dev/narwhals/issues/2634
- [x] stricter `is_*_dataframe` https://github.com/narwhals-dev/narwhals/issues/1437
- [x] deprecate `get_level` from main namespace
- [x] any/all horizontal ignore_nulls #2678 
- [x] `LazyFrame.schema` https://github.com/narwhals-dev/narwhals/issues/2783

not considered blockers, as they don't change backcompat:
- [ ] ~~IO functions https://github.com/narwhals-dev/narwhals/issues/2116~~
- [ ] ~~`keep` argument in `mode` #1793~~

**Comment by MarcoGorelli** [comment]
I think I'd also like to remove row-order-dependent expressions:

<details>
- arg_max
- arg_min
- cum_count
- cum_max
- cum_min
- cum_prod
- cum_sum
- diff
- drop_nulls
- ewm_mean
- filter
- gather_every
- head
- is_first_distinct
- is_last_distinct
- mode
- rolling_mean
- rolling_std
- rolling_sum
- rolling_var
- sample
- shift
- sort
- unique
</details>

These can all stay available for `Series`, so people would still be able to use them for eager execution

All of these would of course stay available in `narwhals.stable.v1`

---

In https://github.com/narwhals-dev/narwhals/pull/1689#issuecomment-2566470490, we find:
- Plotly is using `Expr.cum_sum`
- scikit-lego is using `Expr.shift`

So, it would take more planning to be able to do this safely

Example (maybe?) of how `shift_by` could be implemented for Polars:
```python
def shift_by(expr, by):
    idx = pl.arg_sort_by(by)
    return pl.col(expr).gather(idx).shift().sort_by(idx)
```

**Reply by MarcoGorelli** [reply]
🤔 `fill_null` with strategies like `forward` would also become problematic

it's debatable whether to have `by` kwargs (which would be required for the lazy-only layer) or to have `*_by` variants

**Reply by MarcoGorelli** [reply]
I just tried the following:
```python
import polars as pl
import numpy as np

rng = np.random.default_rng(1)
N = 10_000_000
a = rng.normal(size=N)
b = rng.normal(size=N)
c = rng.normal(size=N)
data = {'a': a, 'b': b, 'c': c}
df_pl = pl.DataFrame(data)
df_pd = pd.DataFrame(data)

def shift_by(expr, by, lag):
    idx = pl.arg_sort_by(by)
    return pl.col(expr).gather(idx).shift(lag).sort_by(idx)

def add_lags_polars_1(df, cols, lags):
    return df.with_columns(
        pl.col(col).shift(-lag).alias(f'{col}_{lag}')
        for col in cols
        for lag in lags
    )

def add_lags_polars_2(df, cols, lags):
    return df.with_row_index().with_columns(
        shift_by(col, 'index', -lag).alias(f'{col}_{lag}')
        for col in cols
        for lag in lags
    )

def add_lags_polars_3(df, cols, lags):
    return df.with_columns(
        df[col].shift(-lag).alias(f'{col}_{lag}')
        for col in cols
        for lag in lags
    )
```
Timings:
- add_lags_polars_1 (current): 1.63ms
- add_lags_polars_2 (using `shift_by`): ~1300ms
- add_lags_polars_3 (using `Series.shift`): 4.3ms

So, I'm not sure it's really warranted to remove `Expr.shift`

**Comment by MarcoGorelli** [comment]
Something else to discuss is `nw.is_ordered_categorical`, whose signature is:
```
 is_ordered_categorical(series) 
```

The signature means we need to have an (eager) Series to determine whether a column is ordered categorical, which is problematic for lazy implementations.

What we could have instead is:
```python
is_ordered_categorical(frame, column_name) -> bool
```
Then, the logic would go:
- pandas, Polars, PyArrow: same as now
- duckDB / Ibis / PySpark (which have no Categorical data type, as far as I can tell?): always return `False`
- as a last fallback, for anything eager which implements the dataframe interchange protocol, check if this data type is described as ordered categorical

**EDIT**

The above change might not actually be necessary. In Altair, they need the unique categories:

https://github.com/vega/altair/blob/be5e9ecd61d099c847dca44fb3a4283940c1e3b8/altair/utils/core.py#L707-L710

So, we could do something like this there:
```python
if (
    # Getting the unique categories would require materialising a LazyFrame,
    # so we only auto-infer ordinal in the eager case.
    isinstance(df, nw.DataFrame)
    and nw.is_ordered_categorical(df[col_name])
    and not (categories := df[col_name].cat.get_categories()).is_empty()
):
    return "ordinal", categories.to_list()
```

**Comment by MarcoGorelli** [comment]
Another step closer to stable.v2: all expressions should either aggregate or preserve length. Anything different isn't generally supported in lazy backends - I think Polars is the only exception here, and that's because of some different design choices. Which is great for anyone using Polars, but anyone trying to use Narwhals to write dataframe-agnostic code should not be doing such operations and expecting them to work

What we might be able to allow is expressions that change length if they're all followed by aggregations. Sql allows this (`sum(a) filter (where b>3)`) but it's not clear to me if duckdbs relational api does. Same with distinct. The rest though we may want to disallow? 

Draft PR: https://github.com/narwhals-dev/narwhals/pull/1791

**Reply by FBruzzesi** [reply]
> What we might be able to allow is expressions that change length if they're all followed by aggregations. Sql allows this (sum(a) filter (where b>3)) but it's not clear to me if duckdbs relational api does. Same with distinct.

- I am quite pro supporting these kind of expression, with a clear disclaimer
- IIRC this would be something similar to what I tried in #743.
- An additional case to allow is a select context with a single expression (although this could fallback to a dataframe method instead).

**Reply by MarcoGorelli** [reply]
Yup, exactly! I think we can do some validation for this at the `nw.Expr` level, instead of within all the compliant frames

Concretely, the ones that we could allow (if eventually followed by an aggregation) would be the ones with a sql equivalent:
- Expr.filter (sql example: `select sum(a) filter where (b > 1) from rel`)
- Expr.drop_nulls (sql example: `select sum(a) filter where a is not null from rel`)
- Expr.unique (sql example: `select sum(distinct a) from rel`)
- Expr.mode (not sure about the sql equivalent right now but maybe there is one)

I think that the index alignment in Dask would make this hard to do in the general case there, so Dask may still need some exceptions...(e.g. `(nw.col('a').drop_nulls() + nw.col('b').drop_nulls()).sum()`...)

The ones that I think we probably shouldn't allow (because they're row-order dependent and have no sql equivalent) are:
- Expr.sort
- Expr.head
- Expr.tail
- Expr.gather_every
- Expr.sample

> An additional case to allow is a select context with a single expression (although this could fallback to a dataframe method instead).

I'm less sure about this one tbh, it might be simpler to teach (and to implement) if the rule is simple: the lazyframes, expressions must either eventually aggregate or preserve length

---

Rules draft:
- changes_length means "change length but doesn't aggregate"
- no n-ary operation may include an expression which changes length
- applying an aggregation to an expression which changes length removes the "changes_length" attribute
- no expression pass to `select` / `with_columns` may change length

**Comment by MarcoGorelli** [comment]
> Later we can add an order_by argument to these functions which will become required when working with nw.Lazyframe

One more comment on this - I think this would be much simpler to implement as
```python
nw.col('sales').cum_sum().over(order_by='date')
```
as opposed to
```python
nw.col('sales').cum_sum(order_by='date')
```

Because then, we just have a single `over` context, and it's straightforward to also allow
```python
nw.col('sales').cum_sum().over('store', order_by='date')
```
whereas `nw.col('sales').cum_sum(order_by='date').over('store')` would require us to fuse two window expressions together. Not saying that can't be done, but I'd prefer to keep things simple

---

Having said that, `nw.col('sales').rolling_mean(2).over('store', order_by='date')` would also require us to fuse `over`s. So, we may need to get https://github.com/narwhals-dev/narwhals/issues/1848 and then look into fusing `over`s

**Reply by MarcoGorelli** [reply]
Some more rules for order-dependent operations:

1. Any order-dependent function must immediately be followed by an `over`, in which `order_by` (and optionally `partition_by`) are specified, e.g.:
   - `nw.col('a').cum_sum().over('store', order_by='date')` is allowed
   - `nw.col('a').cum_sum().abs().over('store', order_by='date')` is not
2. Order-dependent functions can not be called on CHANGES_LENGTH expressions. `nw.col('a').drop_nulls().cum_sum().over('store', order_by='date')` is not allowed

This should allow for a fairly straightforward translation of Narwhals API to PySpark / SQL:
- `cum_sum` just applies `sum`, and with `.over('store', order_by='date')` it becomes `sum(a) over (partition by store order by date rows between unbounded preceding and current row)`
- `rolling_sum(2)` also just applies `sum`, but we store the metadata `2` somewhere, and with `.over('store', order_by='date')` it becomes `case when count(a) over (partition by store order by date rows between 1 preceding and current row) >= 2 then sum(a) over (partition by store order by date rows between 1 preceding and current row) else null end`

---

Note that all of the above only applies to `LazyFrame`s. `DataFrame`s are pretty free from restrictions

---

Some operations which we should aim for. `df` denotes `DataFrame` (eager), `lf` `LazyFrame` (lazy):
- eager chain of order-dependent functions without `over` `df.select(nw.col('a').shift(1).rolling_mean(2))`: allowed
- eager chain of order-dependent functions with order-by `over` `df.select(nw.col('a').shift(1).rolling_mean(2).over(order_by='i'))`: allowed. Not allowed if there's a `partition_by` in `over`, e.g. `.over('b', order_by='i')`, as pandas doesn't allow non-elementary group-bys
- lazy chain of order-dependent operations `lf.select(nw.col('a').shift(1).rolling_mean(2).over(order_by='i'))`: not allowed
- lazy order-dependent operation in combination with elementwise operation `lf.select(nw.col('a').abs().rolling_mean(2).over(order_by='i'))` allowed
- lazy order-dependent operation in combination with n-ary operation of literals or elementwise `lf.select((nw.col('a')+1).rolling_mean(2).over(order_by='i'))` allowed
- n-ary operation inside `over` `lf.select((nw.col('a').abs().rolling_mean(2)+1).over(order_by='i'))` are not allowed. Instead, users can do `lf.select((nw.col('a').abs().rolling_mean(2)).over(order_by='i'))+1`

**Reply by dangotbanned** [reply]
I'm surprised this didn't get mentioned before, but wouldn't [`polars.Expr.sort_by`](https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.sort_by.html) be an alternative to use in some of these cases?

I was looking at [Order-dependence](https://narwhals-dev.github.io/narwhals/basics/order_dependence/) which has this example:

```py
import pandas as pd
from sqlframe.duckdb import DuckDBSession

import narwhals as nw

session = DuckDBSession()
data = {"a": [1, 3, 4], "i": [0, 1, 2]}

df_pd = pd.DataFrame({"a": [1, 3, 4], "i": [0, 1, 2]})
sqlframe_df = session.createDataFrame(df_pd)
ldf = nw.from_native(sqlframe_df)

>>> ldf.with_columns(a_cum_sum=nw.col("a").cum_sum().over(order_by="i")).collect("pandas")
┌──────────────────┐
|Narwhals DataFrame|
|------------------|
|   a  i  a_cum_sum|
|0  1  0          1|
|1  3  1          4|
|2  4  2          8|
└──────────────────┘
```

Translating that `Expr` directly to `polars` isn't valid at runtime:

```py
import polars as pl

ldf = pl.LazyFrame(data)

>>> ldf.with_columns(a_cum_sum=pl.col("a").cum_sum().over(order_by="i")).collect()
TypeError: Expr.over() missing 1 required positional argument: 'partition_by'
```

But we can use the shorter `sort_by`:

```py
>>> ldf.with_columns(a_cum_sum=pl.col("a").cum_sum().sort_by("i")).collect()
shape: (3, 3)
┌─────┬─────┬───────────┐
│ a   ┆ i   ┆ a_cum_sum │
│ --- ┆ --- ┆ ---       │
│ i64 ┆ i64 ┆ i64       │
╞═════╪═════╪═══════════╡
│ 1   ┆ 0   ┆ 1         │
│ 3   ┆ 1   ┆ 4         │
│ 4   ┆ 2   ┆ 8         │
└─────┴─────┴───────────┘
```

@MarcoGorelli I feel like I must be missing something 😄 

The API for `sort_by` matches what we need for determining order-dependence and has all the arguments you'd expect for the SQL-based backends:
- `*by`
- `descending`
- `nulls_last`

Also seems to have been available as far back as [`0.18.0`](https://docs.pola.rs/docs/python/version/0.18/reference/expressions/api/polars.Expr.sort_by.html), but without `nulls_last` support until [`0.20.*`](https://docs.pola.rs/api/python/version/0.20/reference/expressions/api/polars.Expr.sort_by.html)

*Personally*, I quite like how we'd be able to suggest this in an error without mentioning windows/over

**Reply by MarcoGorelli** [reply]
that should be fine too, so long as (for lazy backends) we keep the same restrictions on what `over(order_by=...)` can be called on, else we end up with
```python
df.select(nw.col('a'), nw.col('b').sort_by('i'))
```
where the rows get scrambled

**Reply by dangotbanned** [reply]
Yeah same restrictions was what I was thinking as well!

They might even share a lot of the same implementation.
I'm thinking of it mostly as a more ergonomic and discoverable path to solve the same problem 🙂

`Expr.over` can do a lot of things, whereas `Expr.sort_by` is more narrowly defined for ordering

**Reply by MarcoGorelli** [reply]
Having looked at this more closely, it's not the same, see https://github.com/narwhals-dev/narwhals/issues/2534#issuecomment-2872057450

**Comment by dangotbanned** [comment]
Have you considered moving some of the instance attributes into `ClassVar`(s)?
[Example](https://github.com/narwhals-dev/narwhals/pull/1924#issuecomment-2631115464) of proposing for `nw.Schema`.

Here, I'm thinking more in terms of things like:

https://github.com/narwhals-dev/narwhals/blob/4a2ca52b52a52a6f8fb9d915ad4c347c02b9fbde/narwhals/dataframe.py#L73-L77

Instead you'd have:

```py
class BaseFrame(Generic[_FrameT]):
    _compliant_frame: Any
    _level: ClassVar[Literal["full", "lazy"]]

    @classmethod
    def _from_compliant_dataframe(cls, df: Any):
        # NOTE: `_level` would already be accessible in `__init__`
        return cls(df)
```

Looking at `_stableify`, you might be able to benefit from [`__init_subclass__`](https://docs.python.org/3/reference/datamodel.html#customizing-class-creation). 

Achieves the same goal, but you get a context that is evaluated **once per-class** (e.g. enforce some error handling at declaration-time)

<details><summary><code>_stableify</code></summary>
<p>

https://github.com/narwhals-dev/narwhals/blob/4a2ca52b52a52a6f8fb9d915ad4c347c02b9fbde/narwhals/stable/v1/__init__.py#L1085-L1122

</p>
</details> 

Short example of what that *could* look like:

```py
# nw.dataframe.DataFrame / BaseFrame
class NwDataFrame(Generic[IntoFrameT]):
    def __init_subclass__(
        cls, *args: Any, version: Version = Version.MAIN, **kwds: Any
    ) -> None:
        super().__init_subclass__(*args, **kwds)
        cls._version: Version = version


class DataFrame(NwDataFrame[IntoDataFrameT], version=Version.V1): ...
```

I'm just picking out `version` & `level` as two common arguments.
This *could* apply anywhere that some attribute is fixed for all instances of a class.

Maybe there is good reason to be able to configure this on the instance-level - *but if not* - you'd end up with less context that needs passing around 🙂

**Reply by MarcoGorelli** [reply]
Thanks for looking into this! As far as I understood, this would not be user-facing, right? If so, definitely on-board with refactoring

**Reply by dangotbanned** [reply]
> Thanks for looking into this! As far as I understood, this would not be user-facing, right? If so, definitely on-board with refactoring

Great @MarcoGorelli 

Yeah I don't think any of this would be user-facing.

Making changes like this for `v2` would impact internals and *probably* extension authors.
To what extent though depends on what things you might want to move around

**Reply by MarcoGorelli** [reply]
that's fine then, the extensions mechanism is still marked as experimental (and I don't know of anyone who's actually tried extending Narwhals yet)

**Reply by dangotbanned** [reply]
> Achieves the same goal, but you get a context that is evaluated **once per-class** (e.g. enforce some error handling at declaration-time

One attribute that wouldn't work with this is `backend_version: tuple[int, ...]`.
AFAICT, for that you'd want something more like a *write-once* attribute, **per-implementation**
https://github.com/narwhals-dev/narwhals/blob/f79a9ea2292905b57ea3cfbda15b0f37d1221562/narwhals/utils.py#L387-L398

You can't know what version is installed (if any) at class-definition time. 
However, you would know that the version cannot change after it has been seen once.

So for something like (https://github.com/narwhals-dev/narwhals/pull/2007#issuecomment-2657181479), you can be sure `backend_version` is known after the first `ArrowSeries` is instantiated.

`PandasLike...` would probably need the most thought on if it could support this

**Reply by dangotbanned** [reply]
Using `__init_subclass__` for `Version` in https://github.com/narwhals-dev/narwhals/pull/2324#discussion_r2027396829

**Comment by MarcoGorelli** [comment]
I think we're in a good place, not too long until `stable.v2` I think

We probably really need to make a decision around "empty sum: 0 or null?" though:
- leave it backend-dependent and document the differences
- unify on 0
- unify on null
